packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart

```diff
@@ -0,0 +1,766 @@
+        _DeferredComponentAndroidFiles(
+          name: component.name,
+          env: env,
+          templatesDir: _templatesDir
+        );
+      if (!androidFiles.verifyFilesExist()) {
+        // generate into temp directory
+        final Map<String, List<File>> results =
+          await androidFiles.generateFiles(
+            alternateAndroidDir: _outputDir,
+            clearAlternateOutputDir: true,
+          );
+        for (final File file in results['outputs']) {
+          _generatedFiles.add(file.path);
+        }
+        _outputs.addAll(results['outputs']);
+        _inputs.addAll(results['inputs']);
+      }
+    }
+  }
+
+  // The key used to identify the metadata element as the loading unit id to deferred component mapping.
+  static const String _mappingKey = 'io.flutter.embedding.engine.deferredcomponents.DeferredComponentManager.loadingUnitMapping';
+  // Matches meta-data elements that uses the _mappingKey with any amount of whitespace.
+  static final RegExp _mappingMetadataMatcher = RegExp(r'<meta-data[\s\n]*android:name[\s\n]*=[\s\n]*"' + _mappingKey + r'"');
+  // Matches the beginning of all meta-data elements.
+  static final RegExp _metadataEntryMatcher = RegExp(r'<meta-data');
+  // Matches the closing tag of the application block of an AndroidManfiest.
+  static final RegExp _manifestApplicationEntryCloseMatcher = RegExp(r'^[\s\n]*</application>');
+
+  /// Checks if the base module `app`'s `AndroidManifest.xml` contains the required meta-data
+  /// that maps loading units to deferred components.
+  ///
+  /// Flutter engine uses a manifest metadata mapping to determine what deferred component
+  /// includes a particular loading unit id. This method checks if `app`'s `AndroidManifest.xml`
+  /// contains this metadata. If not, it will generate a modified AndroidManifest.xml with the
+  /// correct metadata entry.
+  void checkAppAndroidManifestComponentLoadingUnitMapping(List<DeferredComponent> components, List<LoadingUnit> generatedLoadingUnits) {
+    final Directory androidDir = env.projectDir.childDirectory('android');
+    List<String> lines;
+
+    // We do not use the Xml package to handle the writing, as we do not want to
+    // erase any user applied formatting and comments. The changes can be
+    // applied with dart io and custom parsing.
+    final File appManifestFile = androidDir
+      .childDirectory('app')
+      .childDirectory('src')
+      .childDirectory('main')
+      .childFile('AndroidManifest.xml');
+    _inputs.add(appManifestFile);
+    XmlDocument document;
+    try {
+      document = XmlDocument.parse(appManifestFile.readAsStringSync());
+    } on XmlParserException {
+      throwToolExit('Error parsing $appManifestFile '
+                    'Please ensure that the android manifest is a valid XML document and try again.');
+    } on FileSystemException {
+      throwToolExit('Error reading $appManifestFile even though it exists. '
```

Comment:
The tool should correctly throw errors for permissions into tool exits, so you might only hit a fse if the file is missing.